xtask/src/semver_check.rs

```diff
@@ -0,0 +1,278 @@
+use std::{
+    fs,
+    path::{Path, PathBuf},
+};
+
+use esp_metadata::Chip;
+use rustdoc_types::ItemEnum;
+
+use crate::{Package, cargo::CargoArgsBuilder};
+
+pub fn generate_baseline(
+    workspace: &PathBuf,
+    packages: Vec<Package>,
+    chips: Vec<Chip>,
+) -> Result<(), anyhow::Error> {
+    for package in packages {
+        log::info!("Generating API baseline for {package}");
+
+        for chip in &chips {
+            log::info!("Chip = {}", chip.to_string());
+            let package_name = package.to_string();
+            let package_path = crate::windows_safe_path(&workspace.join(&package_name));
+
+            let current_path = build_doc_json(package, chip, &package_path)?;
+
+            let suffix = if package.chip_features_matter() {
+                format!("_{}", chip.to_string())
+            } else {
+                "".to_string()
+            };
+
+            let to_path = PathBuf::from(&package_path).join(format!("api_baseline{}.json", suffix));
+
+            remove_unstable_items(&current_path)?;
+
+            log::debug!("Copy {current_path:?}");
+            fs::copy(current_path, to_path)?;
+
+            if !package.chip_features_matter() {
+                break;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub fn check(
+    workspace: &PathBuf,
+    packages: Vec<Package>,
+    chips: Vec<Chip>,
+) -> Result<(), anyhow::Error> {
+    for package in packages {
+        log::info!("Semver-check API for {package}");
+
+        for chip in &chips {
+            log::info!("Chip = {}", chip.to_string());
+            let package_name = package.to_string();
+            let package_path = crate::windows_safe_path(&workspace.join(&package_name));
+
+            let current_path = build_doc_json(package, chip, &package_path)?;
+
+            let suffix = if package.chip_features_matter() {
+                format!("_{}", chip.to_string())
+            } else {
+                "".to_string()
+            };
+
+            let baseline_path = package_path.join(format!("api_baseline{}.json", suffix));
+
+            remove_unstable_items(&current_path)?;
+
+            let mut check = cargo_semver_checks::Check::new(
+                cargo_semver_checks::Rustdoc::from_path(current_path),
+            );
+            check.set_baseline(cargo_semver_checks::Rustdoc::from_path(baseline_path));
+            let mut cfg = cargo_semver_checks::GlobalConfig::new();
+            // seems there is no way of getting the details via the API - so lets log them
+            cfg.set_log_level(Some(log::Level::Info));
+            let check_result = check.check_release(&mut cfg)?;
+
+            log::info!("Result {:?}", check_result);
+
+            for (_, report) in check_result.crate_reports() {
+                if let Some(required_bump) = report.required_bump() {
+                    if required_bump == cargo_semver_checks::ReleaseType::Major {
```

Comment:
Does this return `Major` when bumping pre 1.0 crates? i.e 0.15 -> 0.16

Reply:
from testing this works as expected

Reply:
What happens to pre-releases like `-beta.0`? cargo-semver-checks doesn't seem to have a concept of those.

Reply:
yes - but in our case it would be e.g `1.0.0-beta.0` - `1.0.0-beta.0` and the result would be to do a patch, minor or (hopefully not) major change. That works fine. After bumping we will create new baselines and it's the same then

it would cause problem if we compare a `1.0.0-beta.0` and a `1.0.0-beta.1` - i.e. check if it's correctly bumped - that wouldn't work but it's not the way we use it - we just check what we would need if we would want to release

I hope that explains my thoughts here